### PR TITLE
fix failed test cases and add changes to run using openshift-tuning namespace

### PR DESCRIPTION
**Pull request by khansaad**
This PR fixes issue #81 

Signed-off-by: Saad Khan <saakhan@redhat.com>

### CODE REVIEW

File: deploy_hpo.sh [hunk 1]
```diff
@@ -50,8 +51,15 @@ function usage() {
 	echo " -o: build with specific hpo container image name [Default - kruize/hpo:<version>]"
 	echo " -n: Namespace to which hpo is deployed [Default - monitoring namespace for cluster type minikube]"
 	echo " -d: Config maps directory [Default - manifests/configmaps]"
+	echo " Environment Variables to be set: REG_EMAIL, REG_UNAME, REG_PASS"
 	exit -1
 }
+# check if registry credentials are set as Environment Variables
+if [ -z $REG_UNAME ] || [ -z $REG_PASS ] || [ -z $REG_EMAIL ]; then
```
Comment (chandrams) [inline deploy_hpo.sh:58]: This check should be made only if the cluster type is openshift or minikube

Comment (chandrams) [reply deploy_hpo.sh:58]: Seems to be used only in openshift-helpers script, so the PR checks should work fine on other cluster types, if condition is added to check cluster type as openshift

Comment (khansaad) [reply deploy_hpo.sh]: Added condition for both Minikube and Openshift.

File: deploy_hpo.sh [hunk 2]
```diff
@@ -105,6 +107,15 @@ fi
 SERVICE_STATUS_NATIVE=$(ps -u | grep service.py | grep -v grep)
 SERVICE_STATUS_DOCKER=$(${CONTAINER_RUNTIME} ps | grep hpo_docker_container)
 
+# In case of Minikube and Openshift, check if registry credentials are set as Environment Variables
+if [ ${cluster_type} == "minikube" || ${cluster_type} == "openshift" ]; then
+    if [ -z $REG_UNAME ] || [ -z $REG_PASS ] || [ -z $REG_EMAIL ]; then
```
Comment (dinogun) [inline deploy_hpo.sh:112]: Please have all variables in quotes and brackets "${VAR}"

File: scripts/cluster-helpers.sh [hunk 3]
```diff
@@ -125,6 +125,12 @@ function minikube_first() {
 	kubectl create namespace ${hpo_ns}
 
 	kubectl_cmd="kubectl -n ${hpo_ns}"
+
+    # create a kube secret each time app is deployed
+    kubectl create secret docker-registry hpodockersecret --docker-server=docker.io --docker-username=$REG_UNAME \
+     --docker-password=$REG_PASS --docker-email=$REG_EMAIL -n ${hpo_ns}
+    # link the secret to the service account
+    oc secrets link default hpodockersecret --for=pull -n ${hpo_ns}
```
Comment (dinogun) [inline scripts/cluster-helpers.sh:133]: Looks like space / tab issue

File: scripts/openshift-helpers.sh [hunk 4]
```diff
@@ -23,6 +23,12 @@ function openshift_first() {
 	kubectl create namespace ${hpo_ns}
 
 	kubectl_cmd="kubectl -n ${hpo_ns}"
+
+    # create a kube secret each time app is deployed
+    kubectl create secret docker-registry hpodockersecret --docker-server=docker.io --docker-username=$REG_UNAME \
+     --docker-password=$REG_PASS --docker-email=$REG_EMAIL -n ${hpo_ns}
+    # link the secret to the service account
+    oc secrets link default hpodockersecret --for=pull -n ${hpo_ns}
```
Comment (dinogun) [inline scripts/openshift-helpers.sh:31]: space tab issue here as well. Instead of duplicating this code, please add a new function and move it to common code

File: tests/scripts/hpo/hpo_multiple_exp_tests.sh [hunk 5]
```diff
@@ -240,7 +240,10 @@ function hpo_multiple_exp_test() {
 
 			http_code=$(tail -n1 <<< "${get_trial_json}")
 			response=$(echo -e "${get_trial_json}" | tail -2 | head -1)
-			response=$(echo ${response} | cut -c 4-)
+
+            if [ ${response::3} == "000" ]; then
+                response=$(echo ${response} | cut -c 4-)
+            fi
```
Comment (dinogun) [inline tests/scripts/hpo/hpo_multiple_exp_tests.sh:246]: Why is this needed? Also formatting issues

File: tests/scripts/hpo/hpo_sanity_tests.sh [hunk 6]
```diff
@@ -173,12 +171,15 @@ function hpo_sanity_test() {
...
+		get_trial_json_cmd="${curl} ${hpo_url}?experiment_name="petclinic-sample-2-75884c5549-npvgd"&trial_number=${i} -w '\n%{http_code}'"
 		echo "command used to query the experiment_trial API = ${get_trial_json_cmd}" | tee -a ${LOG}
 
 		http_code=$(tail -n1 <<< "${get_trial_json}")
 		response=$(echo -e "${get_trial_json}" | tail -2 | head -1)
-		response=$(echo ${response} | cut -c 4-)
+
+		if [ ${response::3} == "000" ]; then
+			response=$(echo ${response} | cut -c 4-)
+		fi
```
Comment (dinogun) [inline tests/scripts/hpo/hpo_sanity_tests.sh:182]: Why is this needed?

Comment (khansaad) [reply tests/scripts/hpo/hpo_sanity_tests.sh]: Resolved

File: tests/scripts/hpo/hpo_multiple_exp_tests.sh (hunk 5 above)
Comment (khansaad) [reply tests/scripts/hpo/hpo_multiple_exp_tests.sh]: Resolved

File: scripts/openshift-helpers.sh (hunk 4 above)
Comment (khansaad) [reply scripts/openshift-helpers.sh]: Done

File: scripts/cluster-helpers.sh (hunk 3 above)
Comment (khansaad) [reply scripts/cluster-helpers.sh]: Resolved

File: deploy_hpo.sh (hunk 2 above)
Comment (khansaad) [reply deploy_hpo.sh]: Done

File: tests/scripts/hpo/hpo_multiple_exp_tests.sh [hunk 7]
```diff
@@ -240,7 +240,12 @@ function hpo_multiple_exp_test() {
 
 			http_code=$(tail -n1 <<< "${get_trial_json}")
 			response=$(echo -e "${get_trial_json}" | tail -2 | head -1)
-			response=$(echo ${response} | cut -c 4-)
+
+			# Added condition to check for '000' as sometimes cURL command returns it due to reasons such as
+			# 'Failed DNS resolution','connection refused' or 'timed out'
+			if [ ${response::3} == "000" ]; then
+				response=$(echo ${response} | cut -c 4-)
+			fi
```
Comment (dinogun) [inline tests/scripts/hpo/hpo_multiple_exp_tests.sh:248]: Looks like this is common code that is used in multiple places. Can you create a function for this instead?

File: deploy_hpo.sh [hunk 8]
```diff
@@ -46,6 +47,7 @@ function usage() {
 	echo " -o: build with specific hpo container image name [Default - kruize/hpo:<version>]"
 	echo " -n: Namespace to which hpo is deployed [Default - monitoring namespace for cluster type minikube]"
 	echo " -d: Config maps directory [Default - manifests/configmaps]"
+	echo " Environment Variables to be set: REG_EMAIL, REG_UNAME, REG_PASS"
```
Comment (dinogun) [inline deploy_hpo.sh:50]: I think the variables can be a bit more descriptive, I'd suggest
```
REGISTRY_EMAIL
REGISTRY_USERNAME
REGISTRY_PASSWORD
```

**Comment by dinogun** [comment]
Need to fix conflicts

File: deploy_hpo.sh (hunk 8 above)
Comment (khansaad) [reply deploy_hpo.sh]: Done

File: deploy_hpo.sh [hunk 9]
```diff
@@ -46,6 +47,7 @@ function usage() {
 	echo " -o: build with specific hpo container image name [Default - kruize/hpo:<version>]"
 	echo " -n: Namespace to which hpo is deployed [Default - monitoring namespace for cluster type minikube]"
 	echo " -d: Config maps directory [Default - manifests/configmaps]"
+	echo " Environment Variables to be set: REGISTRY, REGISTRY_EMAIL, REGISTRY_USERNAME, REGISTRY_PASSWORD"
```
Comment (dinogun) [inline deploy_hpo.sh:50]: Give an example for `REGISTRY` as `docker.io`, `quay.io`

**Comment by dinogun** [comment]
Pls fix the conflicts

File: tests/scripts/hpo/hpo_multiple_exp_tests.sh [hunk 10]
```diff
@@ -235,7 +235,9 @@ function hpo_multiple_exp_test() {
 
 			http_code=$(tail -n1 <<< "${get_trial_json}")
 			response=$(echo -e "${get_trial_json}" | tail -2 | head -1)
```
Comment (dinogun) [inline tests/scripts/hpo/hpo_multiple_exp_tests.sh:237]: It looks like this line can be moved into the `curl_check_error` function as well

File: deploy_hpo.sh (hunk 9 above)
Comment (khansaad) [reply deploy_hpo.sh]: Done

File: tests/scripts/hpo/hpo_multiple_exp_tests.sh (hunk 10 above)
Comment (khansaad) [reply tests/scripts/hpo/hpo_multiple_exp_tests.sh]: Moved

File: scripts/cluster-helpers.sh [hunk 11]
```diff
@@ -275,3 +278,14 @@ function check_prereq() {
...
+
+# create kubernetes secret
+function create_secret() {
+
+	namespace="$1"
+	# create a kube secret each time app is deployed
+	kubectl create secret docker-registry hpodockersecret --docker-server=$REGISTRY --docker-username=$REGISTRY_USERNAME \
+	--docker-password=$REGISTRY_PASSWORD --docker-email=$REGISTRY_EMAIL -n ${namespace}
+	# link the secret to the service account
+	oc secrets link default hpodockersecret --for=pull -n ${namespace}
```
Comment (dinogun) [inline scripts/cluster-helpers.sh:290]: It looks like this is used by both minikube and OpenShift. We cannot use the `oc` command for minikube. Please use kubectl for both

File: .github/workflows/test_on_pr.yaml [hunk 12]
```diff
@@ -7,7 +7,11 @@ on:
 
   # Allows you to run this workflow manually from the Actions tab
   workflow_dispatch:
-
+env:
+  REGISTRY: ${{ secrets.REGISTRY }}
+  REGISTRY_USERNAME: ${{ secrets.REGISTRY_USERNAME }}
+  REGISTRY_PASSWORD: ${{ secrets.REGISTRY_PASSWORD }}
+  REGISTRY_EMAIL: ${{ secrets.REGISTRY_EMAIL }}
```
Comment (dinogun) [inline .github/workflows/test_on_pr.yaml:14]: Is this needed now that we have moved minikube out?

File: scripts/cluster-helpers.sh [hunk 13]
```diff
@@ -277,3 +280,20 @@ function check_prereq() {
...
+	# create a kube secret each time app is deployed
+	kubectl create secret docker-registry hpodockersecret --docker-username="${REGISTRY_USERNAME}" \
+	--docker-server="${REGISTRY}" --docker-email="${REGISTRY_EMAIL}"  --docker-password="${REGISTRY_PASSWORD}" \
+	-n ${namespace}
+
+	# link the secret to the service account
+	if [ "${cluster_type}" == "minikube" ]; then
+		kubectl patch serviceaccount default -p '{"imagePullSecrets": [{"name": "hpodockersecret"}]}' -n ${namespace}
+	else
+		oc secrets link default hpodockersecret --for=pull -n ${namespace}
```
Comment (dinogun) [inline scripts/cluster-helpers.sh:297]: As I mentioned previously, we cannot use `oc` in common code. This needs to be kubectl

Comment (khansaad) [reply scripts/cluster-helpers.sh]: Updated now
